fx quant: ensure fp32 sum works correctly with quantized layers

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53875 fx quant: remove unnecessary dequants for getattr
* #53614 fx quant: clean up names of quantize handlers
* **#53196 fx quant: ensure fp32 sum works correctly with quantized layers**
* #53187 fx quant: fix using size of quant layer in torch._assert
* #53120 fx quant: support for ndim followed by binary op

Summary:

Before this PR, code patterns like this did not work:

```
x = some_quant_layer(x)
x = torch.stack([x, ...])
x = torch.sum(x, ...)
```

The reason this did not work is because `torch.sum` is treated as
"quantized" because of the newly added fp16 support, even though it is
not actually "quantized" for models where fp16 is not used.  We may
need to adjust the concept of "quantized vs non-quantized" into a
"dtype" for the longer term fix.

The PR adds logic to skip matching `sum` to `BinaryOp` if the dtype
requested is not supported.

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_quant_sum
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26783960](https://our.internmc.facebook.com/intern/diff/D26783960)

vkuzo committed:
[not for land] fix fx quant for quant_layer -> stack -> sum

Summary:

Before this PR, code patterns like this did not work:

```
x = some_quant_layer(x)
x = torch.stack([x, ...])
x = torch.sum(x, ...)
```

The reason this did not work is because `torch.sum` is treated as
"quantized" because of the newly added fp16 support, even though it is
not actually "quantized" for models where fp16 is not used.  We may
need to adjust the concept of "quantized vs non-quantized" into a
"dtype" for the longer term fix.

The current PR is a hacky fix to unblock.  We need to clean things
up before this is landable

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_quant_sum
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[not for land] fix fx quant for quant_layer -> stack -> sum"


Summary:

Before this PR, code patterns like this did not work:

```
x = some_quant_layer(x)
x = torch.stack([x, ...])
x = torch.sum(x, ...)
```

The reason this did not work is because `torch.sum` is treated as
"quantized" because of the newly added fp16 support, even though it is
not actually "quantized" for models where fp16 is not used.  We may
need to adjust the concept of "quantized vs non-quantized" into a
"dtype" for the longer term fix.

The current PR is a hacky fix to unblock.  We need to clean things
up before this is landable

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_quant_sum
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26783960](https://our.internmc.facebook.com/intern/diff/D26783960)

[ghstack-poisoned]

vkuzo committed:
Update on "[not for land] fix fx quant for quant_layer -> stack -> sum"


Summary:

Before this PR, code patterns like this did not work:

```
x = some_quant_layer(x)
x = torch.stack([x, ...])
x = torch.sum(x, ...)
```

The reason this did not work is because `torch.sum` is treated as
"quantized" because of the newly added fp16 support, even though it is
not actually "quantized" for models where fp16 is not used.  We may
need to adjust the concept of "quantized vs non-quantized" into a
"dtype" for the longer term fix.

The current PR is a hacky fix to unblock.  We need to clean things
up before this is landable

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_quant_sum
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26783960](https://our.internmc.facebook.com/intern/diff/D26783960)

[ghstack-poisoned]

vkuzo committed:
Update on "[not for land] fix fx quant for quant_layer -> stack -> sum"


Summary:

Before this PR, code patterns like this did not work:

```
x = some_quant_layer(x)
x = torch.stack([x, ...])
x = torch.sum(x, ...)
```

The reason this did not work is because `torch.sum` is treated as
"quantized" because of the newly added fp16 support, even though it is
not actually "quantized" for models where fp16 is not used.  We may
need to adjust the concept of "quantized vs non-quantized" into a
"dtype" for the longer term fix.

The current PR is a hacky fix to unblock.  We need to clean things
up before this is landable

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_quant_sum
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26783960](https://our.internmc.facebook.com/intern/diff/D26783960)

[ghstack-poisoned]

vkuzo committed:
Update on "[not for land] fix fx quant for quant_layer -> stack -> sum"


Summary:

Before this PR, code patterns like this did not work:

```
x = some_quant_layer(x)
x = torch.stack([x, ...])
x = torch.sum(x, ...)
```

The reason this did not work is because `torch.sum` is treated as
"quantized" because of the newly added fp16 support, even though it is
not actually "quantized" for models where fp16 is not used.  We may
need to adjust the concept of "quantized vs non-quantized" into a
"dtype" for the longer term fix.

The current PR is a hacky fix to unblock.  We need to clean things
up before this is landable

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_quant_sum
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26783960](https://our.internmc.facebook.com/intern/diff/D26783960)

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -1744,6 +1744,44 @@ def forward(self, x):
+            qconfig_dict = {'': torch.quantization.default_qconfig}
+            mp = prepare_fx(m, qconfig_dict)
+            mp(torch.rand(4, 1, 4, 4))
+            mc = convert_fx(mp)
```
In this case, shouldnt we be inserting an observer at the output of sum?  Also for stack, do we assume that all the tensors stacked have the same scale and zero-point?

vkuzo replied:
> shouldnt we be inserting an observer at the output of sum?

yes, I believe this is implemented as "at the input of conv" instead of "at the output of sum", even though the result is the same

> for stack, do we assume that all the tensors stacked have the same scale and zero-point?

quantized stack currently uses the qparams of the first input for the output.  That could be definitely be improved.

jerryzh168 commented on torch/quantization/fx/quantize.py:
```diff
@@ -1050,19 +1050,28 @@ def record_match(pattern, node, matched):
-                        # break after finding the first match
-                        break
+
+                            this_node_qconfig = self.qconfig_map[node.name]
```
here we are coupling matching with quantize handler I think

vkuzo replied:
Correct.  It looks like the current matching strategy of a match being determined only by the node type and nothing else does not scale to some of these new requirements (this PR, and the PR earlier in this stack).  We now need to match by node type, dtype, and the existence of tensors (versus primitives) in the node's args.

My understanding of the high level concepts:
* `QuantizeHandler`: determines how a node or a fusion pattern should go through prepare and convert steps
* matching: matches tuples of nodes to corresponding quantize handler instances

One way to clean this up in the future is to remove the restriction that a single pattern can only match to a single instance of quantize handler, and make the matching process aware of this and do the match correctly with the extra information.

jerryzh168 replied:
I think it makes sense to extend matching to consider qconfig as well, but currently we don't do that. sounds good to me if we plan to clean this up later.

jerryzh168 replied:
maybe add some comments to say that we'll extend matching to consider qconfigs and clean this up?

vkuzo committed:
Update on "fx quant: ensure fp32 sum works correctly with quantized layers"


Summary:

Before this PR, code patterns like this did not work:

```
x = some_quant_layer(x)
x = torch.stack([x, ...])
x = torch.sum(x, ...)
```

The reason this did not work is because `torch.sum` is treated as
"quantized" because of the newly added fp16 support, even though it is
not actually "quantized" for models where fp16 is not used.  We may
need to adjust the concept of "quantized vs non-quantized" into a
"dtype" for the longer term fix.

The PR adds logic to skip matching `sum` to `BinaryOp` if the dtype
requested is not supported.

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_quant_sum
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26783960](https://our.internmc.facebook.com/intern/diff/D26783960)

[ghstack-poisoned]

vkuzo committed:
Update on "fx quant: ensure fp32 sum works correctly with quantized layers"


Summary:

Before this PR, code patterns like this did not work:

```
x = some_quant_layer(x)
x = torch.stack([x, ...])
x = torch.sum(x, ...)
```

The reason this did not work is because `torch.sum` is treated as
"quantized" because of the newly added fp16 support, even though it is
not actually "quantized" for models where fp16 is not used.  We may
need to adjust the concept of "quantized vs non-quantized" into a
"dtype" for the longer term fix.

The PR adds logic to skip matching `sum` to `BinaryOp` if the dtype
requested is not supported.

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_quant_sum
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26783960](https://our.internmc.facebook.com/intern/diff/D26783960)

[ghstack-poisoned]
